src/main/java/com/google/api/codegen/discovery/transformer/csharp/CSharpSampleTypeNameConverter.java

```diff
@@ -71,33 +75,59 @@ public TypeName getServiceTypeName(String apiTypeName) {
 
   @Override
   public TypeName getRequestTypeName(String apiTypeName, TypeInfo typeInfo) {
-    return getTypeName(typeInfo);
+    List<String> copy = new ArrayList<>(methodNameComponents);
+    String requestTypeName = copy.remove(copy.size() - 1) + "Request";
+    String path = "";
+    for (String s : copy) {
+      path += s + "Resource.";
+    }
+    return new TypeName(path + requestTypeName);
   }
 
   @Override
   public TypeName getTypeName(TypeInfo typeInfo) {
-    if (typeInfo.isMessage()) {
-      return new TypeName(typeInfo.message().typeName());
+    if (typeInfo.isMap()) {
+      TypeName mapTypeName = typeNameConverter.getTypeName("System.Collections.Generic.Dictionary");
+      TypeName keyTypeName = getTypeName(typeInfo.mapKey());
+      TypeName valueTypeName = getTypeName(typeInfo.mapValue());
+      return new TypeName(
+          mapTypeName.getFullName(),
+          mapTypeName.getNickname(),
+          "%s<%i, %i>",
+          keyTypeName,
+          valueTypeName);
     }
-    return getNonMessageTypeName(typeInfo);
+    if (typeInfo.isArray()) {
+      TypeName listTypeName = typeNameConverter.getTypeName("System.Collections.Generic.List");
+      TypeName elementTypeName = getTypeNameForElementType(typeInfo);
```

Comment:
Don't we just need a field like `typeInfo.arrayElement()` analogous to `typeInfo.mapKey()` and `typeInfo.mapValue()`? It feels bad that all these cases were supported back in the `DiscoveryContext` era but not anymore. While I don't recall yet seeing arrays of arrays, I think I have seen maps of maps.

I guess tracking the issue is fine for now.